In [1]:
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv('Day10_Flight_Operations_Dataset.csv')

# Display the first few rows to understand the structure
display(df.head())

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5


In [2]:
print("--- Dataset Information ---")
df.info()

print("\n--- Summary Statistics ---")
display(df.describe(include='all'))

print("\n--- Missing Values Check ---")
print(df.isnull().sum())

--- Dataset Information ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               180 non-null    object 
 1   Flight_Date             180 non-null    object 
 2   Airline                 180 non-null    object 
 3   Origin                  180 non-null    object 
 4   Destination             180 non-null    object 
 5   Aircraft                180 non-null    object 
 6   Travel_Class            180 non-null    object 
 7   Passengers              180 non-null    int64  
 8   Seat_Capacity           180 non-null    int64  
 9   Average_Ticket_Price    180 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           180 non-null    object 
 12  Weather                 180 non-null    object 
 13  Booking_Channel         180 non-null    object 
 14  Avg_Baggage_Kg

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
count,180,180,180,180,180,180,180,180.000000,180.000000,180.000000,173.000000,180,180,180,180.000000,180,180.000000
unique,180,111,6,11,8,4,3,NaN,NaN,NaN,NaN,3,5,4,NaN,4,NaN
top,FL0001,2026-02-10,Vistara,Delhi,Delhi,Airbus A319,Economy,NaN,NaN,NaN,NaN,On Time,Rain,Online Travel Portal,NaN,Vegetarian,NaN
freq,1,6,37,43,51,60,134,NaN,NaN,NaN,NaN,114,43,50,NaN,52,NaN
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,122.772222,197.238889,6111.677778,19.335260,NaN,NaN,NaN,15.150000,NaN,3.494444
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,45.635651,15.408176,3615.220040,29.029712,NaN,NaN,NaN,5.705506,NaN,1.169908
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,45.000000,180.000000,1949.000000,0.000000,NaN,NaN,NaN,5.000000,NaN,2.000000
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,78.750000,186.000000,3815.500000,0.000000,NaN,NaN,NaN,10.000000,NaN,2.000000
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,125.000000,189.000000,5380.000000,5.000000,NaN,NaN,NaN,15.500000,NaN,3.000000
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,162.250000,210.000000,6790.000000,20.000000,NaN,NaN,NaN,20.000000,NaN,5.000000



--- Missing Values Check ---
Flight_ID                 0
Flight_Date               0
Airline                   0
Origin                    0
Destination               0
Aircraft                  0
Travel_Class              0
Passengers                0
Seat_Capacity             0
Average_Ticket_Price      0
Delay_Minutes             7
Flight_Status             0
Weather                   0
Booking_Channel           0
Avg_Baggage_Kg            0
Meal_Preference           0
Passenger_Satisfaction    0
dtype: int64


In [3]:
# 1. Fill missing values in 'Delay_Minutes' with 0 (Assuming NaN means no delay)
df['Delay_Minutes'] = df['Delay_Minutes'].fillna(0)

# 2. Create a new column 'Total_Revenue' (Passengers * Average_Ticket_Price)
df['Total_Revenue'] = df['Passengers'] * df['Average_Ticket_Price']

# 3. Create a 'Route' column by concatenating Origin and Destination
df['Route'] = df['Origin'] + " to " + df['Destination']

# 4. Calculate 'Load_Factor' percentage (Passengers / Seat_Capacity)
df['Load_Factor'] = (df['Passengers'] / df['Seat_Capacity']) * 100

print("Data cleaning and transformations applied. New columns added.")
display(df[['Flight_ID', 'Route', 'Total_Revenue', 'Load_Factor']].head())

Data cleaning and transformations applied. New columns added.


,Flight_ID,Route,Total_Revenue,Load_Factor
0,FL0001,Mumbai to Bengaluru,1447740,116.666667
1,FL0002,Delhi to Mumbai,312760,33.333333
2,FL0003,Kochi to Delhi,315771,30.454545
3,FL0004,Srinagar to Delhi,580023,69.545455
4,FL0005,Pune to Delhi,1157890,98.888889


In [4]:
# Filter: Identify significantly delayed flights (Delay > 30 minutes)
significantly_delayed = df[df['Delay_Minutes'] > 30]
print(f"Total flights delayed by more than 30 minutes: {len(significantly_delayed)}")

# Sort: Top 5 most delayed flights
top_delayed = df.sort_values(by='Delay_Minutes', ascending=False).head(5)
print("\n--- Top 5 Most Delayed Flights ---")
display(top_delayed[['Flight_ID', 'Airline', 'Route', 'Delay_Minutes', 'Weather', 'Flight_Status']])

# Filter & Sort: Cancelled flights sorted by date
cancelled_flights = df[df['Flight_Status'] == 'Cancelled'].sort_values(by='Flight_Date')
print(f"\n--- Cancelled Flights ({len(cancelled_flights)}) ---")
display(cancelled_flights[['Flight_Date', 'Airline', 'Route', 'Weather']])

Total flights delayed by more than 30 minutes: 40

--- Top 5 Most Delayed Flights ---


,Flight_ID,Airline,Route,Delay_Minutes,Weather,Flight_Status
50,FL0051,Air India,Delhi to Mumbai,110.0,Rain,Delayed
26,FL0027,SpiceJet,Bengaluru to Hyderabad,110.0,Clear,Delayed
82,FL0083,Air India Express,Lucknow to Delhi,110.0,Cloudy,Delayed
57,FL0058,IndiGo,Hyderabad to Kolkata,110.0,Cloudy,Delayed
177,FL0178,IndiGo,Delhi to Mumbai,110.0,Fog,Delayed



--- Cancelled Flights (7) ---


,Flight_Date,Airline,Route,Weather
65,2026-03-06,SpiceJet,Srinagar to Delhi,Cloudy
68,2026-03-09,SpiceJet,Kochi to Delhi,Cloudy
40,2026-03-23,Akasa Air,Delhi to Chennai,Fog
37,2026-03-27,Air India,Delhi to Srinagar,Rain
160,2026-04-03,Vistara,Delhi to Mumbai,Storm
110,2026-05-16,Vistara,Delhi to Mumbai,Cloudy
23,2026-05-19,SpiceJet,Ahmedabad to Bengaluru,Rain


In [5]:
# Group by Airline: Calculate total flights, passengers, revenue, and average satisfaction
airline_performance = df.groupby('Airline').agg(
    Total_Flights=('Flight_ID', 'count'),
    Total_Passengers=('Passengers', 'sum'),
    Average_Load_Factor=('Load_Factor', 'mean'),
    Total_Revenue=('Total_Revenue', 'sum'),
    Avg_Satisfaction=('Passenger_Satisfaction', 'mean')
).sort_values(by='Total_Revenue', ascending=False)

print("\n--- Airline Performance Overview ---")
display(airline_performance)

# Group by Weather: See how weather impacts average delays
weather_delays = df.groupby('Weather')['Delay_Minutes'].mean().reset_index().sort_values(by='Delay_Minutes', ascending=False)

print("\n--- Impact of Weather on Delays ---")
display(weather_delays)

# Group by Route: Most popular routes by flight frequency
popular_routes = df['Route'].value_counts().reset_index().head(5)
popular_routes.columns = ['Route', 'Flight_Count']
print("\n--- Top 5 Most Popular Routes ---")
display(popular_routes)


--- Airline Performance Overview ---


,Total_Flights,Total_Passengers,Average_Load_Factor,Total_Revenue,Avg_Satisfaction
Airline,,,,,
Vistara,37,4484,61.868716,28689467,3.378378
IndiGo,32,4158,66.778509,25709131,3.343750
SpiceJet,32,3929,63.604102,24195935,3.531250
Air India Express,24,3027,64.786071,23944164,3.666667
Akasa Air,31,3427,55.212934,18872059,3.419355
Air India,24,3074,64.655403,16232418,3.750000



--- Impact of Weather on Delays ---


,Weather,Delay_Minutes
3,Rain,21.186047
4,Storm,19.433333
1,Cloudy,18.447368
0,Clear,18.029412
2,Fog,15.342857



--- Top 5 Most Popular Routes ---


,Route,Flight_Count
0,Delhi to Chennai,17
1,Mumbai to Bengaluru,16
2,Srinagar to Delhi,15
3,Kochi to Delhi,14
4,Delhi to Mumbai,13


# Key Observations & Findings

*   Vistara leads the market in total flights (37) and total revenue (₹28,689,467), followed closely by IndiGo which carried the most passengers overall (4,158).

*   Interestingly, Air India records the highest average passenger satisfaction score (3.75 out of 5), despite operating fewer flights (24) compared to Vistara and IndiGo.

*   The vast majority of flights operate normally, with 114 out of 180 flights documented as "On Time". Only 7 flights in the entire dataset were "Cancelled".

*   As expected, weather plays a role in flight punctuality. Flights operating in Rain experienced the highest average delay times (21.2 minutes), followed by Storms (19.4 minutes). Surprisingly, Fog had the lowest average delay profile (15.3 minutes).

*   The highest trafficked routes clearly connect major metropolitan hubs. Delhi to Chennai (17 flights) and Mumbai to Bengaluru (16 flights) are the top two most frequently flown routes in this dataset.

*   IndiGo operates with the highest average Load Factor (approx. 66.78%), meaning they are consistently filling more of their aircraft's seat capacity compared to competitors like Akasa Air, which averages the lowest load factor (55.21%).

*   The dataset contains 180 entries, with a high degree of completeness. Only the Delay_Minutes column had missing values (7 records), which were logically filled with 0s indicating no delay occurred.

